File: src/main/java/ro/nicolaemariusghergu/queryit/service/CategoryService.java
package ro.nicolaemariusghergu.queryit.service;

import org.springframework.lang.NonNull;
import ro.nicolaemariusghergu.queryit.model.Category;

import java.util.List;
import java.util.Optional;

public interface CategoryService {

    @NonNull
    Optional<Category> findById(@NonNull Long id);

    @NonNull
    List<Category> findAll();

    Category update(Category category);

    Optional<Category> findByName(String name);

    <S extends Category> S save(S entity);

    <S extends Category> List<S> saveAll(Iterable<S> entities);

    @NonNull
    void deleteById(@NonNull Long id);
}

File: src/main/kotlin/ro/nicolaemariusghergu/queryit/service/CategoryService.kt
package ro.nicolaemariusghergu.queryit.service

import org.springframework.http.ResponseEntity
import org.springframework.lang.NonNull
import ro.nicolaemariusghergu.queryit.dto.CategoryDto


interface CategoryService {
    fun getCategoryById(@NonNull id: Long): ResponseEntity<CategoryDto>

    fun getCategories(): ResponseEntity<List<CategoryDto>>

    fun updateCategory(category: CategoryDto): ResponseEntity<CategoryDto>

    fun getCategoryByName(name: String): ResponseEntity<CategoryDto>

    fun addCategory(categoryDto: CategoryDto): ResponseEntity<Long>

    fun deleteCategoryById(@NonNull id: Long): ResponseEntity<Long>
}